File: aeds2/verde/arvore-binaria/ArvoreBinaria.java
import java.util.*;

/*
*Faça a implementação de uma árvore binária de pesquisa (ABP) em Java para armazenar números inteiros. Crie um programa que instancia uma nova árvore e processa as operações solicitadas na entrada de dados, obedecendo o seguinte formato:
*'I X': insere o número X;
*'P Y': pesquisa se o número Y está na árvore, imprimindo S caso verdadeiro e N caso contrário. O algoritmo de pesquisa deve imprimir cada nó visitado;
*'PRE': imprime os elementos da árvore por caminhamento pré-ordem ou V se a árvore estiver vazia;
*'POS': imprime os elementos da árvore por caminhamento pós-ordem ou V se a árvore estiver vazia;
*'EM': imprime os elementos da árvore por caminhamento em-ordem ou V se a árvore estiver vazia;
*/

public class ArvoreBinaria{
	public static void main(String[] args){
		Scanner sc = new Scanner(System.in);
		Arvore a = new Arvore();
		String comando;

		while(sc.hasNext()){
			comando = sc.nextLine();//pesquisar diferença entre funcoes next

			if(comando.equals("I")){
				int elementoInserir = sc.nextInt();
				System.out.println(comando);
				System.out.println(elementoInserir);
				a.inserirElemento(elementoInserir);
			}
			else if(comando.equals("EM")){
				a.exibirEmOrdem();
			}
		}
	}
}

class No{
	private int elemento;
	private No direita;
	private No esquerda;

	public No(int elemento){
		this.elemento = elemento;
		this.direita = null;
		this.esquerda = null;
	}

	public int getElemento(){
		return this.elemento;
	}

	public No getDireita(){
		return this.direita;
	}

	public void setDireita(No direita){
		this.direita = direita;
	}

	public No getEsquerda(){
		return this.esquerda;
	}

	public void setEsquerda(No esquerda){
		this.esquerda = esquerda;
	}


}

class Arvore{
	private No raiz;

	public Arvore(){//Cria a estrutura inicial e inicializa a raiz como "null"
		this.raiz = null;
	}
	
	public void inserirElemento(int elementoInserir){
		this.raiz = inserirRecursivo(elementoInserir, this.raiz);
	}
	
	private No inserirRecursivo(int elementoInserir, No i){//Private por que é chamado somente pelo método "inserirElemento"
		if(i == null){
			i = new No(elementoInserir);
		}
		else if(elementoInserir > i.getElemento()){
			i.setDireita(inserirRecursivo(elementoInserir, i.getDireita()));
		}
		else{
			i.setEsquerda(inserirRecursivo(elementoInserir, i.getEsquerda()));
		}

		return i;
	}

	public void exibirEmOrdem(){
		emOrdemRecursivo(this.raiz);
	}

	private void emOrdemRecursivo(No i){
		if(i != null){
			emOrdemRecursivo(i.getEsquerda());
			System.out.println(i.getElemento());
			emOrdemRecursivo(i.getDireita());
		}
		else{
			System.out.println("V");
		}
	}
}

File: aeds2/verde/arvore-binaria/c/ArvoreBinaria.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

struct No {
	int elemento;
	struct No *esq;
	struct No *dir;
};

struct ArvoreBinaria{
	struct No* raiz;
};

// =========================
// Inserir
// =========================
struct No* inserirRec(struct No* i, int elementoInserir){
	if(i == NULL){
		struct No* novo = (struct No*) malloc(sizeof(struct No));
		novo->elemento = elementoInserir;
		novo->esq = NULL;
		novo->dir = NULL;

		return novo;
	}
	else if(i->elemento > elementoInserir){
		i->esq = inserirRec(i->esq, elementoInserir);
	}
	else{
		i->dir = inserirRec(i->dir, elementoInserir);
	}

	return i;
}

void inserir(struct ArvoreBinaria* arvore, int elementoInserir){
	arvore->raiz = inserirRec(arvore->raiz, elementoInserir);
}

// =========================
// Pesquisar
// =========================
int pesquisarRec(struct No* i, int elementoPesquisar){
	int resp = 0;

	if(i == NULL){
		resp = 0;
	}
	else if(i->elemento == elementoPesquisar){
		printf("%d ", i->elemento);
		resp = 1;
	}
	else if(i->elemento > elementoPesquisar){
		printf("%d ", i->elemento);
		resp = pesquisarRec(i->esq, elementoPesquisar);
	}
	else{
		printf("%d ", i->elemento);
		resp = pesquisarRec(i->dir, elementoPesquisar);
	}

	
	return resp;
}

void pesquisar(struct ArvoreBinaria* arvore, int elementoPesquisar){
	printf("%s\n", pesquisarRec(arvore->raiz, elementoPesquisar) ? "S" : "N");
}

// =========================
// Caminhar Pré-ordem
// =========================
void caminharPreRec(struct No* i){
	if(i != NULL){
		printf("%d ", i->elemento);
		caminharPreRec(i->esq);
		caminharPreRec(i->dir);
	}
}

void caminharPreOrdem(struct ArvoreBinaria* arvore){
	if(arvore->raiz != NULL){
		caminharPreRec(arvore->raiz);
	}
	else{
		printf("V");
	}
}

// =========================
// Caminhar Pós-ordem
// =========================
void caminharPosRec(struct No* i){
	if(i != NULL){
		caminharPosRec(i->esq);
		caminharPosRec(i->dir);
		printf("%d ", i->elemento);
	}
}

void caminharPosOrdem(struct ArvoreBinaria* arvore){
	if(arvore->raiz != NULL){
		caminharPosRec(arvore->raiz);
	}
	else{
		printf("V");
	}
}

// =========================
// Caminhar Em-ordem
// =========================
void caminharCentralRec(struct No* i){
	if(i != NULL){
		caminharCentralRec(i->esq);
		printf("%d ", i->elemento);
		caminharCentralRec(i->dir);
	}
}

void caminharCentral(struct ArvoreBinaria* arvore){
	if(arvore->raiz != NULL){
		caminharCentralRec(arvore->raiz);
	}
	else{
		printf("V");
	}
}

// =========================
// Main
// =========================
int main(){
	struct ArvoreBinaria arvore;
	char comando[10];
	
	while(scanf("%s", comando) != EOF){
	
		if(strcmp(comando, "I") == 0){
			int elementoInserir;
			scanf("%d", &elementoInserir);
			inserir(&arvore, elementoInserir);
		}
		else if(strcmp(comando, "EM") == 0){
			caminharCentral(&arvore);
			printf("\n");
		}
		else if(strcmp(comando, "PRE") == 0){
			caminharPreOrdem(&arvore);
			printf("\n");
		}
		else if(strcmp(comando, "POS") == 0){
			caminharPosOrdem(&arvore);
			printf("\n");
		}
		else if(strcmp(comando, "P") == 0){
			int elementoPesquisar;
			scanf("%d", &elementoPesquisar);
			pesquisar(&arvore, elementoPesquisar);
		}
	}
}
